### PR TITLE
[fix] (Nereids) fix date function rewrite on datetimev1 column bug

### LINKED ISSUE #32329: [Bug] [INVALID_ARGUMENT]Create Expr failed because [E33] Invalid value: 2024-03-11 24:00:00 for type DateTime
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

2.0.5-rc02
2.1.0-rc11

### What's Wrong?

With nereids planner enabled, `date()` function will be rewritten by the `DateFunctionRewrite` class.

<br />

But when a datetimev1 column is processed by `date()` function then compares to a date literal, typically equal to and less than equal, it goes wrong due to **padding the date literal with time '24:00:00'**, this datetime literal cannot pass range check when it comes to BE, because hour in a datetime literal **must not be greater than 23**

<br />

Code location:

org.apache.doris.nereids.trees.expressions.literal.DateLiteral
```java
    public DateTimeLiteral toEndOfTheDay() {
        return new DateTimeLiteral(year, month, day, 24, 0, 0);
    }
```

be/src/vec/runtime/vdatetime_value.cpp
```cpp
bool VecDateTimeValue::check_range(uint32_t year, uint32_t month, uint32_t day, uint32_t hour,
                                   uint32_t minute, uint32_t second, uint16_t type) {
    bool time = hour > (type == TIME_TIME ? TIME_MAX_HOUR : 23) || minute > 59 || second > 59;
    if (type == TIME_TIME) {
        return time;
    } else {
        return time || check_date(year, month, day);
    }
}
```

### What You Expected?

datetimev1 column works well in nereids planner

### How to Reproduce?

Create a table with a datetimev1 column, use `date()` function to process it, then compare to a date literal.
Only equal to and less than equal can reproduce the bug.

<br />

```sql
-- add "disable_datev1 = false" to fe.conf first
create table table_for_bug_test (
  id int not null comment 'ID',
  datetimev1_column datetimev1 null comment '日期时间v1列'
) ENGINE=OLAP
DUPLICATE KEY (`id`)
DISTRIBUTED BY HASH(`id`) BUCKETS 1
PROPERTIES (
  "replication_num" = "1"
)
;

select * from table_for_bug_test where date(datetimev1_column) = '2024-03-08';

select * from table_for_bug_test where date(datetimev1_column) <= '2024-03-08';
```

### Anything Else?

if datetimev1 and decimalv2 are gonna be deleted in the future, there should be a statement or tool for us to migrate those outdated column type, or we can only execute `ALTER TABLE MODIFY COLUMN` one by one.

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ziyanTOP**
## Proposed changes

Issue Number: close #32329 

Avoid '24:00:00' time literal for datetimev1 column after handling by date function rewrite in nereids planner

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java [hunk 1]
```diff
@@ -458,10 +458,10 @@ public DateTimeLiteral toBeginOfTheDay() {
     /**
      * 2020-01-01
      *
-     * @return 2020-01-01 24:00:00
+     * @return 2020-01-01 23:59:59
      */
     public DateTimeLiteral toEndOfTheDay() {
-        return new DateTimeLiteral(year, month, day, 24, 0, 0);
+        return new DateTimeLiteral(year, month, day, 23, 59, 59);
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java:464]: please add a sql based regression test

Comment (ziyanTOP) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java:464]: done

**Comment by ziyanTOP** [comment]
run buildall

**Review by 924060929 — APPROVED** [review]

**Comment by 924060929** [comment]
@ziyanTOP good job👍🏻, would you submit a fix pr for branch-2.0?

**Review by morrySnow — APPROVED** [review]

**Comment by ziyanTOP** [comment]
> @ziyanTOP good job👍🏻, would you submit a fix pr for branch-2.0?

ok, I will

**Comment by ziyanTOP** [comment]
done @924060929

**Review by yiguolei — APPROVED** [review]

**Pull request merged**